taichi/ir/frontend_ir.h

```diff
@@ -408,11 +408,31 @@ class ExternalTensorExpression : public Expression {
   int element_dim;  // 0: scalar; 1: vector (SOA); 2: matrix (SOA); -1: vector
                     // (AOS); -2: matrix (AOS)
 
+  // Fill element shape if compile-time specialization is desired.
+  std::vector<int> element_shape;
+
   ExternalTensorExpression(const DataType &dt,
                            int dim,
                            int arg_id,
                            int element_dim)
-      : dt(dt), dim(dim), arg_id(arg_id), element_dim(element_dim) {
+      : dt(dt),
+        dim(dim),
+        arg_id(arg_id),
+        element_dim(element_dim),
+        element_shape() {
+    set_attribute("dim", std::to_string(dim));
+  }
+
+  ExternalTensorExpression(const DataType &dt,
+                           int dim,
+                           int arg_id,
+                           int element_dim,
+                           const std::vector<int> &element_shape)
+      : dt(dt),
+        dim(dim),
+        arg_id(arg_id),
+        element_dim(element_dim),
+        element_shape(element_shape) {
     set_attribute("dim", std::to_string(dim));
   }
 
```

Comment:
@ailzhang What do you think about these two constructors here? I just forgot to merge them :)

Reply:
a cleaner way to write the second is 
```
...
: ExternalTensorExpression(dt, dim, arg_id, element_dim), element_shape(element_shape) {}
```

Reply:
Btw I probably put my comment in the wrong place :P Here's the one I care about https://github.com/taichi-dev/taichi/blob/master/taichi/program/program.cpp#L432

Reply:
Then let's keep two constructors.